financial_calculations/mbs_cash_flows.py

```diff
@@ -0,0 +1,286 @@
+import numpy as np
+import pandas as pd
+from datetime import datetime, timedelta
+from dateutil.relativedelta import relativedelta
+from utils import (
+    discount_cash_flows
+)
+
+def calculate_monthly_payment(principal, num_months, annual_interest_rate):
+    """
+    Calculate the monthly payment for a loan.
+
+    Parameters:
+    principal (float): Principal amount.
+    num_months (int): Number of months.
+    annual_interest_rate (float): Annual interest rate (as a decimal).
+
+    Returns:
+    float: The monthly payment
+
+    Raises:
+    ValueError: If `num_months` is less than or equal to 0.
+    ValueError: If `principal` is less than or equal to 0.
+
+    Notes:
+    - If the `annual_interest_rate` is 0, the monthly payment is simply the principal divided by the number of months.
+    """
+    if num_months <= 0:
+        raise ValueError("Number of months must be greater than zero.")
+    if principal <= 0:
+        raise ValueError("Principal must be greater than zero.")
+    
+    monthly_interest_rate = annual_interest_rate / 12
+    
+    if monthly_interest_rate == 0:
+        return principal / num_months
+
+    payment = (principal * monthly_interest_rate * (1 + monthly_interest_rate) ** num_months) / \
+              ((1 + monthly_interest_rate) ** num_months - 1)
+    
+    return payment
+
+def calculate_scheduled_balances(principal, num_months, annual_interest_rate, monthly_payment):
+    """
+    Calculate scheduled balances, principal paydowns, and interest paid.
+
+    Parameters:
+    principal (float): Principal amount.
+    num_months (int): Number of months.
+    annual_interest_rate (float): Annual interest rate (as a decimal).
+    monthly_payment (float): Monthly payment.
+
+    Returns:
+    tuple: (months, balances, principal_paydowns, interest_paid)
+    """
+    months = np.arange(num_months + 1)
+    balances = np.zeros(num_months + 1)
+    principal_paydowns = np.zeros(num_months + 1)
+    interest_paid = np.zeros(num_months + 1)
+
+    balances[0] = principal
+
+    for month in range(1, num_months + 1):
+        interest_paid[month] = balances[month - 1] * annual_interest_rate / 12
+        principal_paydowns[month] = monthly_payment - interest_paid[month]
+        balances[month] = balances[month - 1] - principal_paydowns[month]
+
+    return months, balances, principal_paydowns, interest_paid
+
+def calculate_scheduled_balances_with_service_fee(principal, num_months, annual_interest_rate, monthly_payment, service_fee_rate):
```

Comment:
this should be much simpler.  call your existing scheduled balance code, and then calculate the net interest on that balance vector.  i.e. (gross - service fee )* balance

Reply:
you made a lot of progress here.  looks good.  I want to ask you a few questions:

why compute interest and prin cash flows in compute scheduled balance?  we almost never use them.  

why do we make the user pass the payment into compute compute scheduled balance?  there are some good reasons, but curious about your thinking.  the code could just call compute payment itself

why not compute principal pay down as the difference between balances?  seems easier when you get into actual balances which are the things that actually drive the cash flows

Reply:
1) because thats the output asked for in the colab, also could be useful in the future, and is get information to plot

2) I did this mainly because I thought there were some scenarios where we might have a monthly payment that is different from our standard computer. Not sure how it works in the real world but i.e. buying this security from someone else who gave buyers different terms. 

3) Mainly because 1) I like how the principal paydown is clearly defined so someone looking at the code understands how it's calculated beyond just the difference in balances and 2) the pay its set up now its super easy to vectorize the pooling factors so I can apply them to scheduled balances with no for loop

Reply:
Great answers!